### ISSUE TITLE
Provide environment variables in package.json for --user-data-dir

### ISSUE THREAD

**Issue by anjanapaulose** [nwjs/nw.js#5363]
Hi,

We have to support nw app , in a terminal server which supports multiple sessions for same user. This is through the registry setting HKEY_LOCAL_MACHINE\ System\ CurrentControlSet\ Control\ TerminalServer\ fSingleSessionPerUser - change the value from 1 to 0.

But in this case nw app is launched only in the first session.  

We can provide -user-data-dir attribute to launch multiple instances. But we want to provide an environment variable to this. If we try this in command prompt as  "nw.exe --user-data-dir=%APPDATA%\\test", this works. But  the same wont work if we use it from package.json as chromium args.

Is there  any way by which we can provide environment variables in package.json

thanks
Anjana

**Comment by anjanapaulose** [comment]
We want to give environment variable %APPDATA%\\%SESSIONNAME%.  Since the sessionname key is unique to user in multiple sessions. This will ensure that a new nw app is opened in a new session, but will show existing one in already opened sessions.

**Comment by ghostoy** [comment]
No `--user-data-dir` is needed. You can listen to `open` event on `nw.App` in your first instance. It is fired when a new instance of nw is opened.

**Comment by offshore** [comment]
@ghostoy, Anjana means some different thing, let me clarify the usage case.
Consider Windows Server which allows multiple people to connect via RDP using single account. So when one starts nw that uses `%APPDATA%/appname` as data directory; this prevents other sessions of same user to run their copy of nw. The solution is to use `%APPDATA%/%SESSIONNAME%/appname`, so within session nw remains single-instanced while allowing other users connected with same account to run their own nw; but essentially the problem is that environment variables in "chromium-args" are not resolved.

@anjanapaulose, maybe the (temporary?) solution for you is to use variables in policy definition, take a look at https://www.chromium.org/administrators/policy-list-3/user-data-directory-variables (however remember that it's not a variables for command line arguments).

SO. My suggestion is to provide an ability to use substitutions in "chromium-args" manifest field just like it's already done for `user-agent` field. List of substitutions that should be supported can be taken from link above. Also I think it's a good idea to provide some more, like %package_dir to be able to put data directory relative to package.json location.

**Comment by anjanapaulose** [comment]
We have not used any chrome specific things as of now. 
We have created another launcher exe, which will open nw with arguments (just like from command prompt). This way we could pass environment variables.

**Comment by anjanapaulose** [comment]
One more issue we faced is that, at first we tried with %sessionname%. This will work, but in case of applications like mRemoteNG, if you are already logged in and, just do a full screen, this will result in a new session but with same sessionname, which will result in opening yet another nw. This will result in two nw apps running in original session.
We found that sessionid is unique in this case also, use ended up using %sessionid%.

Another problem is the cleaning up of all these folders on exit as the sessionid is not guaranteed to be same on next login. So we used the same launcher exe to cleanup any folders without a lockfile, once the app is started(this will help to reuse the existing folder, in case sessionid is same).

**Comment by ghostoy** [comment]
I think I understand your requests: you want NW.js to extract environment variables in `chromium-args` field in `package.json`. Specifically in your case, it's used for `--user-data-dir` to:

1. share same nw instance among users with same `%sessionname%`
2. use different nw instances for users with different `%sessionname%`

Possible workaround I suggest before we start to implement this feature. You can write a batch file as the entrance of your app. In the batch file, compose the `--user-data-dir` with the environment variable and launch nw with it.

**Comment by anjanapaulose** [comment]
Thankyou,  I am already doing it, as mentioned above. I have another exe which launches zapp with  sessionid.

But it would be good to have if nw itself can handle it.

**Comment by offshore** [comment]
@ghostoy, after all I reckon that parsing OS environment vars directly inside `chromium-args` seems legit, it may be more sophisticated to implement that it seems to; I think it's worth to implement separate manifest field `user-data-dir` with fixed substitution list (this should be simpler and even safer).
There are some more path-related chromium cli args: `data-path`, `disk-cache-dir`, `homedir`, `load-extension`, `profile-directory`, `register-font-files`, `allow-file-access` and others; all of them may need further investigation.
